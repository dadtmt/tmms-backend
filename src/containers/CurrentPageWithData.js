import R from 'ramda'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'

import { pageEditorId } from '../config'
import makeReducer from '../reducers/makeReducer'
import Editor from '../components/Editor'

export const getCurrentPageQuery = gql`
query GetPageEditor($pageEditorId: ID!) {
  getPageEditor(id: $pageEditorId) {
    id
    currentPage {
      id
      text
      choices {
        edges {
          node {
            id
            text
            made
          }
        }
      }
    }
  }
}
`

// NOT WORKING reducer, bad hack by refetchQueries in mutation

const mutationHandlers = {
  CreatePage: (state, { createPage: { changedPage } }) => R.set(
    R.lensPath(['getPageEditor', 'currentPage']),
    changedPage,
    state
  )
}

const handlers = {
  APOLLO_MUTATION_RESULT: (state, {
    operationName,
    result: { data }
  }) => R.propOr(
    R.identity,
    operationName,
    mutationHandlers
  )(state, data)
}

export const reducer = makeReducer(handlers)

const emptyPage = {
  choices: [],
  id: '',
  text: ''
}

export const makeProps = R.ifElse(
  R.prop('loading'),
  R.pipe(
    R.pick(['loading']),
    R.merge(emptyPage)
  ),
  R.pipe(
    R.pathOr(emptyPage, ['getPageEditor', 'currentPage']),
    R.over(
      R.lensProp('choices'),
      R.propOr([], 'edges')
    )
  )
)

export default graphql(
  getCurrentPageQuery,
  {
    options: {
      // reducer,
      variables: { pageEditorId }
    }
  }
)(Editor)
