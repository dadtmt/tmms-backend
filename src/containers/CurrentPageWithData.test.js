import { makeProps, reducer } from './CurrentPageWithData'

describe('makeProps', () => {
  it('return empty currentpage on loading', () => {
    const data = { loading: true }
    expect(makeProps(data)).toMatchSnapshot()
  })
  it('return empty currentpage when no currentPage', () => {
    const data = {
      getPageEditor: {
        currentPage: null
      },
      loading: false
    }
    expect(makeProps(data)).toMatchSnapshot()
  })
  it('return currentpage', () => {
    const data = {
      getPageEditor: {
        currentPage: {
          choices: {
            edges: []
          },
          id: 'UGFnZTo3NQ==',
          text: 'some text'
        },
        id: 'UGFnZUVkaXRvcjoy'
      },
      loading: false
    }
    expect(makeProps(data)).toMatchSnapshot()
  })
})

describe('reducer', () => {
  it('update currentPage on createPage mutation', () => {
    const action = {
      operationName: 'CreatePage',
      result: {
        data: {
          createPage: {
            changedPage: {
              choices: {
                edges: []
              },
              id: 'SOME_ID',
              text: 'some text'
            }
          }
        }
      },
      type: 'APOLLO_MUTATION_RESULT'
    }
    const state = {
      getPageEditor: {
        currentPage: {
          choices: {
            edges: []
          }
        }
      }
    }
    expect(
      reducer(state, action)
    ).toMatchSnapshot()
  })
})
