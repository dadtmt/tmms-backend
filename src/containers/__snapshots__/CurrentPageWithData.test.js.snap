exports[`makeProps return currentpage 1`] = `
Object {
  "choices": Array [],
  "id": "UGFnZTo3NQ==",
  "text": "some text",
}
`;

exports[`makeProps return empty currentpage on loading 1`] = `
Object {
  "choices": Array [],
  "id": "",
  "loading": true,
  "text": "",
}
`;

exports[`makeProps return empty currentpage when no currentPage 1`] = `
Object {
  "choices": Array [],
  "id": "",
  "text": "",
}
`;

exports[`reducer update currentPage on createPage mutation 1`] = `
Object {
  "getPageEditor": Object {
    "currentPage": Object {
      "choices": Object {
        "edges": Array [],
      },
      "id": "SOME_ID",
      "text": "some text",
    },
  },
}
`;
